pkg/providers/pricing/pricing.go

```diff
@@ -29,62 +43,189 @@ type Provider interface {
+
+	types := make([]string, len(p.onDemandPrices))
+	for t := range p.onDemandPrices {
+		types = append(types, t)
+	}
+	return types
 }
 
 func (p *DefaultProvider) OnDemandPrice(instanceType string) (float64, bool) {
-	// Implement me
 	p.muOnDemand.RLock()
 	defer p.muOnDemand.RUnlock()
-	return -1.0, false
+
+	price, ok := p.onDemandPrices[instanceType]
+	return price, ok
 }
 
+// Zone parameter is ignored, cause in GCP prices are regional
 func (p *DefaultProvider) SpotPrice(instanceType string, zone string) (float64, bool) {
-	// Implement me
 	p.muSpot.RLock()
 	defer p.muSpot.RUnlock()
-	return -1.0, false
+
+	price, ok := p.spotPrices[instanceType]
+	return price, ok
 }
 
-func (p *DefaultProvider) UpdateOnDemandPricing(ctx context.Context) error {
-	// Implement me
-	p.muOnDemand.RLock()
-	defer p.muOnDemand.RUnlock()
+func (p *DefaultProvider) downloadCSV(ctx context.Context) ([][]string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, pricingCSVURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating request: %w", err)
+	}
+
+	// Download the CSV file
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("downloading CSV: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Read all CSV data
+	reader := csv.NewReader(resp.Body)
+	reader.Comma = ','
+	records, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("reading CSV: %w", err)
+	}
+
+	if len(records) < 2 {
+		return nil, fmt.Errorf("CSV file is empty or has no data")
+	}
+
+	return records, nil
+}
+
+func (p *DefaultProvider) updatePrices(ctx context.Context, priceColumn string, storage *pricesStorage) error {
+	records, err := p.downloadCSV(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Find required columns
+	header := records[0]
+	var priceCol, regionCol, nameCol int
+	for i, col := range header {
+		switch col {
+		case priceColumn:
+			priceCol = i
+		case "region":
+			regionCol = i
+		case "name":
+			nameCol = i
+		}
+	}
+
+	// Process the data
+	newPrices := make(pricesStorage)
+
+	for _, record := range records[1:] {
+		// Skip records that don't match our region
+		if record[regionCol] != p.region {
+			continue
+		}
+
+		machineType := record[nameCol]
+		priceStr := record[priceCol]
+
+		price, err := strconv.ParseFloat(priceStr, 64)
+		if err != nil {
+			klog.Errorf("Error parsing price for %s: %v", machineType, err)
+			continue
+		}
+
+		newPrices[machineType] = price
+	}
+
+	if len(newPrices) == 0 {
+		return fmt.Errorf("no prices retrieved during update")
+	}
+
+	*storage = newPrices
 	return nil
 }
 
+func (p *DefaultProvider) UpdateOnDemandPricing(ctx context.Context) error {
+	p.muOnDemand.Lock()
+	defer p.muOnDemand.Unlock()
+
+	return p.updatePrices(ctx, "hour", &p.onDemandPrices)
+}
+
 func (p *DefaultProvider) UpdateSpotPricing(ctx context.Context) error {
-	// Implement me
-	p.muSpot.RLock()
-	defer p.muSpot.RUnlock()
-	return nil
+	p.muSpot.Lock()
+	defer p.muSpot.Unlock()
```

Comment:
We need a controller to call this function: https://github.com/cloudpilot-ai/karpenter-provider-alibabacloud/blob/e92dee8aa5e8de7c9b2e1268238b8b163a1eb310/pkg/controllers/providers/pricing/controllers.go#L49

Reply:
Implemented provider registration